### PR TITLE
Experimental Spark adapter

### LINKED ISSUE #97: Spark adapter
Causally ordered stream processing with Spark Streaming.

### PR DESCRIPTION
**Pull request by krasserm**
- batch processing (from Cassandra storage backends)
- stream processing (from any storage backend)
- closes #97

### CODE REVIEW

File: eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkBatchAdapter.scala [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *               Relevant storage backend configuration parameters are:
+ *               - `eventuate.log.cassandra.keyspace`: keyspace of the event log table
+ *               - `eventuate.log.cassandra.table-prefix`: prefix of the event log table
+ */
+class SparkBatchAdapter(val context: SparkContext, val config: Config) {
+  private val cassandraSettings =
+    new CassandraEventLogSettings(config)
+
+  private implicit val converter: DurableEventConverter =
+    new DurableEventConverter(ConfigFactory.parseString("akka.remote.netty.tcp.port = 0").withFallback(config))
```
Comment (volkerstampa) [inline eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkBatchAdapter.scala]: How about disabling remoting? `provider = "akka.actor.LocalActorRefProvider"`

File: eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+   */
+  def eventStream(id: String, host: String, port: Int, logName: String, sysName: String = ReplicationConnection.DefaultRemoteSystemName, fromSequenceNr: Long = 1L, storageLevel: StorageLevel = StorageLevel.MEMORY_ONLY): DStream[DurableEvent] =
+    context.receiverStream(new DurableEventReceiver(id, ReplicationConnection(host, port, sysName), logName, fromSequenceNr, storageLevel, config))
+}
+
+private class DurableEventReceiver(id: String, connection: ReplicationConnection, logName: String, fromSequenceNr: Long, storageLevel: StorageLevel, config: Config) extends Receiver[DurableEvent](storageLevel) {
+  @volatile private var system: Option[ActorSystem] = None
+
+  override def onStart(): Unit = {
+    system = Some(ActorSystem("receiver", ConfigFactory.parseString("akka.remote.netty.tcp.port = 0").withFallback(config)))
```
Comment (volkerstampa) [inline eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala]: Same here .. disabling? `provider = "akka.actor.LocalActorRefProvider"`

File: eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    private val acceptor = remoteAcceptor
+    private var acceptorRequestSchedule: Option[Cancellable] = None
+
+    private var connected = false
+
+    def receive = {
+      case GetReplicationEndpointInfoSuccess(info) if !connected =>
+        if (info.logNames.contains(logName)) {
+          val sourceLogId = info.logId(logName)
+          val source = ReplicationSource(info.endpointId, logName, sourceLogId, remoteAcceptor)
```
Comment (volkerstampa) [inline eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala]: Why does this not reference the member `acceptor`?

File: eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLogWriter.scala [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.rbmhtechnology.eventuate._
+
+import scala.collection.immutable.Seq
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util._
+
+private object EventLogWriter {
+  class EventLogWriterActor(val id: String, val eventLog: ActorRef) extends EventsourcedActor {
```
Comment (volkerstampa) [inline eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLogWriter.scala]: Something very similar exists here: `com.rbmhtechnology.eventuate.DeleteEventsSpecLeveldb.Emitter`. Maybe it makes sense to reuse and drop one or the other.

**Comment by volkerstampa** [comment]
LGTM

File: eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkBatchAdapter.scala (hunk 1 above)
Comment (krasserm) [inline eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkBatchAdapter.scala]: +1, that's a better solution for the batch adapter.

File: eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala (hunk 3 above)
Comment (krasserm) [inline eventuate-adapter-spark/src/main/scala/com/rbmhtechnology/eventuate/adapter/spark/SparkStreamAdapter.scala]: It should, thanks for the hint

File: eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLogWriter.scala (hunk 4 above)
Comment (krasserm) [inline eventuate-core/src/main/scala/com/rbmhtechnology/eventuate/log/EventLogWriter.scala]: +1, will use the new `EventLogWriter` there

**Comment by krasserm** [comment]
Thanks for reviewing @volkerstampa

**Pull request merged**